File: EVSDKDemo/Player/EVDemoWatchRecordViewController.m
//
//  EVDemoWatchRecordViewController.m
//  EVSDKDemo
//
//  Created by Lcrnice on 17/5/12.
//  Copyright © 2017年 cloudfocous. All rights reserved.
//

#import "EVDemoWatchRecordViewController.h"
//#import "EVPlayer.h"
//#import "EVPlayerConfig.h"
#import <EVMediaFramework/EVPlayer.h>
#import <EVMediaFramework/EVPlayerConfig.h>
#import "EVProgressView.h"

@interface EVDemoWatchRecordViewController () <EVPlayerDelegate>

@property (nonatomic, strong) EVPlayer *player;     /**< 播放器 */

@end

@implementation EVDemoWatchRecordViewController {
    NSTimer *_timer;
    UIButton *_statusBtn;
    EVProgressView *_progressView;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    
    [self setUpUIAndConifgPlayer];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

#pragma mark - player
- (void)setUpUIAndConifgPlayer{
    // UI
    CGFloat width = [UIScreen mainScreen].bounds.size.width;
    CGFloat height = [UIScreen mainScreen].bounds.size.height;
    
    _statusBtn = [UIButton buttonWithType:UIButtonTypeCustom];
    _statusBtn.frame = CGRectMake(10, height - 60, 40, 30);
    [_statusBtn setTitle:@"播放" forState:UIControlStateSelected];
    [_statusBtn setTitle:@"暂停" forState:UIControlStateNormal];
    [_statusBtn addTarget:self action:@selector(changeVideoStatus:) forControlEvents:UIControlEventTouchUpInside];
    _statusBtn.backgroundColor = [UIColor colorWithRed:.5f green:.5f blue:.5f alpha:1.f];
    [self.view addSubview:_statusBtn];
    
    _progressView = [[EVProgressView alloc] init];
    _progressView.frame = CGRectMake(_statusBtn.frame.origin.x + _statusBtn.frame.size.width + 10,
                                     height - 60,
                                     width - 70,
                                     30);
    [self.view addSubview:_progressView];
    
    UIView *containerView = [[UIView alloc] initWithFrame:self.view.frame];
    [self.view insertSubview:containerView atIndex:0];
    
    
    // player
    self.player = [[EVPlayer alloc] init];
    self.player.playerContainerView = containerView;
    self.player.currentPlaybackTime = self.watchStartPosition;
    self.player.live = NO;
    self.player.lid = self.vid;
    self.player.delegate = self;
    
    __weak typeof(self) wSelf = self;
    [self.player playPrepareComplete:^(EVPlayerResponseCode responseCode, NSDictionary *result, NSError *err) {
        __strong typeof(wSelf) sSelf = wSelf;
        if (responseCode == EVPlayerResponse_Okay) {
            [sSelf.player play];
        } else {
            NSLog(@"录播播放失败：%@", err);
        }
    }];
    
    // start timer
    [self startTimer];
    
    
    __weak typeof(_player) weakPlayer = _player;
    _progressView.dragingSliderCallback = ^(EVProgressView *progressView, float progress){
        typeof(weakPlayer) strongPlayer = weakPlayer;
        double seekPos = progress * strongPlayer.duration;
        // 使用 currentPlaybackTime 设置为依靠关键帧定位
//        strongPlayer.currentPlaybackTime = seekPos;
        // 使用 seekTo: 为精准定位
        [strongPlayer seekTo:seekPos];
    };
}

- (void)changeVideoStatus:(UIButton *)btn {
    if (_statusBtn.selected) {
        [self.player play];
    }
    else {
        [self.player pause];
        
    }
    
    _statusBtn.selected = !_statusBtn.selected;
}

#pragma mark - EVPlayerDelegate
- (void)EVPlayerFirstVideoFrameDidRender:(EVPlayer *)player {
    NSLog(@"视频第一帧已渲染");
}
- (void)EVPlayerDidFinishPlay:(EVPlayer *)player reason:(MPMovieFinishReason)reason {
    NSLog(@"播放完毕");
}

- (void)EVPlayer:(EVPlayer *)player cacheState:(EVPlayerCacheState)state {
    NSLog(@"缓冲状态：%@", @(state));
}


#pragma mark - timer
- (void)startTimer {
    if(_timer != nil){
        return;
    }
    _timer = [NSTimer scheduledTimerWithTimeInterval:1.0 target:self selector:@selector(updateState:) userInfo:nil repeats:YES];
}

- (void)stopTimer {
    if (nil == _timer) {
        return;
    }
    [_timer invalidate];
    _timer = nil;
}

- (void)updateState:(NSTimer *)timer {
    if (nil == _player) {
        return;
    }
    
    // 更新播放器的缓冲条
    CGFloat duration = _player.duration;
    CGFloat playableDuration = _player.playableDuration;
    if(duration > 0){
        _progressView.cacheProgress = playableDuration / duration;
    }
    else{
        _progressView.cacheProgress = 0.0;
    }
    
    // 更新进度
    _progressView.totalTimeInSeconds = _player.duration;
    _progressView.playProgress = _player.currentPlaybackTime / _player.duration;
}

#pragma mark - btn actions
- (IBAction)close:(id)sender {
    if (_player) {
        [_player shutDown];
        _player = nil;
    }
    
    [self stopTimer];
    [self dismissViewControllerAnimated:YES completion:nil];
}

- (IBAction)resizePlayer:(id)sender {
    CGFloat width = [UIScreen mainScreen].bounds.size.width;
    CGFloat height = [UIScreen mainScreen].bounds.size.height;
    
    CGRect halfFrame = CGRectMake(width/2, 0, width/2, height/2);
    CGRect originFrame = CGRectMake(0, 0, width, height);
    
    static BOOL changed;
    if (changed == false) {
        self.player.playerViewFrame = halfFrame;
    }
    else {
        self.player.playerViewFrame = originFrame;
    }
    changed = !changed;
}




@end

File: EVSDKDemo_Swift/Player/EVDemoWatchRecordViewController.swift
//
//  EVDemoWatchRecordViewController.swift
//  EVSDKDemo
//
//  Created by Lcrnice on 2017/8/10.
//  Copyright © 2017年 cloudfocous. All rights reserved.
//

import UIKit
import EVMediaFramework

class EVDemoWatchRecordViewController: UIViewController {
  
  var lid = ""
  fileprivate var player: EVPlayer?
  fileprivate var timer: Timer?
  fileprivate var stateBtn: UIButton?
  fileprivate var progressView: EVProgressView?
  fileprivate var containerView: UIView?
  fileprivate let width = UIScreen.main.bounds.width
  fileprivate let height = UIScreen.main.bounds.height
  
  override func viewDidLoad() {
    super.viewDidLoad()
    
    self.configUI()
    self.setupPlayer()
  }
  
  override func didReceiveMemoryWarning() {
    super.didReceiveMemoryWarning()
  }
  
  @IBAction func close(_ sender: Any) {
    if player != nil {
      player?.shutDown()
      player = nil
    }
    
    self.stopTimer()
    self.dismiss(animated: true, completion: nil)
  }
  
  @IBAction func resizePlayer(_ sender: UIButton) {
    let halfFrame = CGRect(x: width/2, y: 0, width: width/2, height: height/2)
    let originFrame = CGRect(x: 0, y: 0, width: width, height: height)
    
    sender.isSelected = !sender.isSelected
    
    if sender.isSelected {
      player?.playerViewFrame = halfFrame
    } else {
      player?.playerViewFrame = originFrame
    }
  }
  
  private func configUI() {
    
    stateBtn = UIButton.init(type: .custom)
    stateBtn?.frame = CGRect(x: 10, y: height - 60, width: 40, height: 30)
    stateBtn?.setTitle("播放", for: .selected)
    stateBtn?.setTitle("暂停", for: .normal)
    stateBtn?.addTarget(self, action: #selector(stateBtnAction(_:)), for: .touchUpInside)
    stateBtn?.backgroundColor = UIColor.lightGray
    view.addSubview(stateBtn!)
    
    progressView = EVProgressView()
    progressView?.frame = CGRect(x: (stateBtn?.frame.origin.x)! + (stateBtn?.frame.width)! + 10, y: height - 60, width: width - 70, height: 30)
    view.addSubview(progressView!)
    
    containerView = UIView(frame: view.frame)
    view.insertSubview(containerView!, at: 0)
  }
  
  private func setupPlayer() {
    player = EVPlayer()
    player?.playerContainerView = containerView
    player?.live = false
    player?.lid = lid
    player?.delegate = self as EVPlayerDelegate
    
    player?.playPrepareComplete({ [weak self] (code: EVPlayerResponseCode, info: [AnyHashable : Any]?, error: Error?) in
      guard let sSelf = self else {
        return
      }
      
      if code == .okay {
        sSelf.player?.play()
      } else {
        print("录播播放失败:\(String(describing: error?.localizedDescription))")
      }
    })
    
    self.startTimer()
    
    progressView?.dragingSliderCallback = { [weak self] (progress: Float) in
      guard let sSelf = self, let sPlayer = sSelf.player else {
        return
      }
      
      let seekPos = Double(progress) * sPlayer.duration
      sPlayer.seek(to: seekPos)
    }
  }
  
  @objc private func stateBtnAction(_ btn: UIButton) {
    if btn.isSelected {
      player?.play()
    } else {
      player?.pause()
    }
    
    btn.isSelected = !btn.isSelected
  }
  
}

// MARK: - Timer
extension EVDemoWatchRecordViewController {
  fileprivate func startTimer() {
    guard timer == nil else {
      return
    }
    
    timer = Timer.scheduledTimer(timeInterval: 1.0, target: self, selector: #selector(updateState(_:)), userInfo: nil, repeats: true)
  }
  
  fileprivate func stopTimer() {
    guard timer != nil else {
      return
    }
    
    timer?.invalidate()
    timer = nil
  }
  
  @objc private func updateState(_ timer: Timer) {
    guard player != nil else {
      return
    }
    
    let duration: Float = Float((player?.duration)!)
    let playableDuration: Float = Float((player?.playableDuration)!)
    
    if duration > 0 {
      progressView?.cacheProgress = playableDuration / duration
    } else {
      progressView?.cacheProgress = 0.0
    }
    
    progressView?.totalTimeInSeconds = duration
    progressView?.playProgress = Float((player?.currentPlaybackTime)! / (player?.duration)!)
  }
}

// MARK: - EVPlayerDelegate
extension EVDemoWatchRecordViewController: EVPlayerDelegate {
  func evPlayerFirstVideoFrameDidRender(_ player: EVPlayer!) {
    print("第一帧已渲染")
  }
  
  func evPlayerDidFinishPlay(_ player: EVPlayer!, reason: MPMovieFinishReason) {
    print("播放完毕")
  }
  
  func evPlayer(_ player: EVPlayer!, cacheState state: EVPlayerCacheState) {
    print("缓冲状态:\(state)")
  }
}
